### PR TITLE
Add table aws_sfn_state_machine closes #701

### LINKED ISSUE #701: Add table aws_stepfunctions_statemachines
**References**
Add any related links that will help us understand the resource, including vendor documentation, related GitHub issues, and Go SDK documentation.

https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-state-machine-structure.html

https://docs.aws.amazon.com/step-functions/latest/apireference/API_ListStateMachines.html

I'd like to be able to query a list of State Machines from the Step Functions service (name, type, creation date, status, arn etc.)

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_sfn_state_machine []

PRETEST: tests/aws_sfn_state_machine

TEST: tests/aws_sfn_state_machine
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_partition.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_iam_role.sfn_role: Creating...
aws_iam_role.sfn_role: Creation complete after 6s [id=turbottest17310]
aws_sfn_state_machine.named_test_resource: Creating...
aws_sfn_state_machine.named_test_resource: Creation complete after 5s [id=arn:aws:states:us-east-1:632902152528:stateMachine:turbottest17310]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

creation_date = 2021-11-02 08:44:23
resource_aka = arn:aws:states:us-east-1:632902152528:stateMachine:turbottest17310
resource_name = turbottest17310
resource_status = ACTIVE

Running SQL query: test-get-call-query.sql
[
  {
    "akas": [
      "arn:aws:states:us-east-1:632902152528:stateMachine:turbottest17310"
    ],
    "name": "turbottest17310",
    "status": "ACTIVE"
  }
]
✔ PASSED

Running SQL query: test-list-call-query.sql
[
  {
    "akas": [
      "arn:aws:states:us-east-1:632902152528:stateMachine:turbottest17310"
    ],
    "name": "turbottest17310",
    "title": "turbottest17310"
  }
]
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "arn:aws:states:us-east-1:632902152528:stateMachine:turbottest17310"
    ],
    "title": "turbottest17310"
  }
]
✔ PASSED

POSTTEST: tests/aws_sfn_state_machine

TEARDOWN: tests/aws_sfn_state_machine

SUMMARY:

1/1 passed.


```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
### Basic info

```sql
select
  name,
  arn,
  status,
  type,
  role_arn
from
  aws_sfn_state_machine;
```

```
+------------+---------------------------------------------------------------+--------+----------+------------------------------------------------------------------------------------+
| name       | arn                                                           | status | type     | role_arn                                                                           |
+------------+---------------------------------------------------------------+--------+----------+------------------------------------------------------------------------------------+
| HelloWorld | arn:aws:states:us-east-1:632902152528:stateMachine:HelloWorld | ACTIVE | STANDARD | arn:aws:iam::632902152528:role/service-role/StepFunctions-HelloWorld-role-4c24fb7f |
+------------+---------------------------------------------------------------+--------+----------+------------------------------------------------------------------------------------+

```

### List active state machines

```sql
select
  name,
  arn,
  status
from
  aws_sfn_state_machine
where
  status = 'ACTIVE';
```

```
+------------+---------------------------------------------------------------+--------+
| name       | arn                                                           | status |
+------------+---------------------------------------------------------------+--------+
| HelloWorld | arn:aws:states:us-east-1:632902152528:stateMachine:HelloWorld | ACTIVE |
+------------+---------------------------------------------------------------+--------+
```

</details>

### CODE REVIEW

File: aws/table_aws_sfn_state_machine.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+				Hydrate:     getAwsStepFunctionsStateMachine,
+			},
+			{
+				Name:        "type",
+				Description: "The type of the state machine.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "creation_date",
+				Description: "The Elasticsearch domain endpoint that use to submit index and search requests.",
```
Comment (rajlearner17) [inline aws/table_aws_sfn_state_machine.go:51]: Is this description right here for Elasticsearch?

File: aws/table_aws_sfn_state_machine.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+			{
+				Name:        "definition",
+				Description: "The Amazon States Language definition of the state machine.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getAwsStepFunctionsStateMachine,
+			},
+			{
+				Name:        "role_arn",
+				Description: "The domain creation status.",
+				Type:        proto.ColumnType_STRING,
```
Comment (rajlearner17) [inline aws/table_aws_sfn_state_machine.go:64]: Should it be `The domain creation status`

File: aws/table_aws_sfn_state_machine.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+	// Build the params
+	params := &sfn.DescribeStateMachineInput{
+		StateMachineArn: &arn,
+	}
+
+	// Get call
+	data, err := svc.DescribeStateMachine(params)
+	if err != nil {
+		logger.Debug("getAwsStepFunctionsStateMachine", "ERROR", err)
```
Comment (rajlearner17) [inline aws/table_aws_sfn_state_machine.go:147]:
```suggestion
		logger.Error("getAwsStepFunctionsStateMachine", "ERROR", err)
```

File: aws/plugin.go [hunk 4]
```diff
@@ -236,6 +236,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"aws_securityhub_hub":                                          tableAwsSecurityHub(ctx),
 			"aws_securityhub_product":                                      tableAwsSecurityhubProduct(ctx),
 			"aws_securityhub_standards_subscription":                       tableAwsSecurityHubStandardsSubscription(ctx),
+			"aws_sfn_state_machine":                                        tableAwsStepFunctionsStateMachine(ctx),
```
Comment (bigdatasourav) [inline aws/plugin.go:239]:
```suggestion
			"aws_sfn_state_machine":                                        tableAwsStepFunctionStateMachine(ctx),
```

File: aws/service.go [hunk 5]
```diff
@@ -1510,6 +1511,29 @@ func SSOAdminService(ctx context.Context, d *plugin.QueryData) (*ssoadmin.SSOAdm
 	return svc, nil
 }
 
+// StepFunctionService returns the service connection for AWS Step Function service
+func StepFunctionService(ctx context.Context, d *plugin.QueryData) (*sfn.SFN, error) {
+	region := d.KeyColumnQualString(matrixKeyRegion)
+	if region == "" {
+		return nil, fmt.Errorf("region must be passed SSOAdminService")
```
Comment (bigdatasourav) [inline aws/service.go:1518]:
```suggestion
		return nil, fmt.Errorf("region must be passed StepFunctionService")
```

File: aws/table_aws_sfn_state_machine.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	"context"
+
+	"github.com/aws/aws-sdk-go/service/sfn"
+
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+)
+
+func tableAwsStepFunctionsStateMachine(_ context.Context) *plugin.Table {
```
Comment (bigdatasourav) [inline aws/table_aws_sfn_state_machine.go:13]:
```suggestion
func tableAwsStepFunctionStateMachine(_ context.Context) *plugin.Table {
```

File: aws/table_aws_sfn_state_machine.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	return &plugin.Table{
+		Name:        "aws_sfn_state_machine",
+		Description: "AWS Step function State Machine",
```
Comment (bigdatasourav) [inline aws/table_aws_sfn_state_machine.go:16]:
```suggestion
		Description: "AWS Step Function State Machine",
```

File: aws/service.go [hunk 8]
```diff
@@ -1510,6 +1511,29 @@ func SSOAdminService(ctx context.Context, d *plugin.QueryData) (*ssoadmin.SSOAdm
 	return svc, nil
 }
 
+// StepFunctionService returns the service connection for AWS Step Function service
+func StepFunctionService(ctx context.Context, d *plugin.QueryData) (*sfn.SFN, error) {
```
Comment (bigdatasourav) [inline aws/service.go:1515]:
```suggestion
func StepFunctionsService(ctx context.Context, d *plugin.QueryData) (*sfn.SFN, error) {
```

File: aws/service.go [hunk 9]
```diff
@@ -1510,6 +1511,29 @@ func SSOAdminService(ctx context.Context, d *plugin.QueryData) (*ssoadmin.SSOAdm
 	return svc, nil
 }
 
+// StepFunctionService returns the service connection for AWS Step Function service
```
Comment (bigdatasourav) [inline aws/service.go:1514]:
```suggestion
// StepFunctionsService returns the service connection for AWS Step Functions service
```

File: aws/service.go [hunk 10]
```diff
@@ -1510,6 +1511,29 @@ func SSOAdminService(ctx context.Context, d *plugin.QueryData) (*ssoadmin.SSOAdm
...
+func StepFunctionService(ctx context.Context, d *plugin.QueryData) (*sfn.SFN, error) {
+	region := d.KeyColumnQualString(matrixKeyRegion)
+	if region == "" {
+		return nil, fmt.Errorf("region must be passed SSOAdminService")
+	}
+	// have we already created and cached the service?
+	serviceCacheKey := fmt.Sprintf("stepfunction-%s", region)
```
Comment (bigdatasourav) [inline aws/service.go:1521]:
```suggestion
	serviceCacheKey := fmt.Sprintf("stepfunctions-%s", region)
```

File: aws/table_aws_sfn_state_machine.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.SingleColumn("arn"),
+			ShouldIgnoreError: isNotFoundError([]string{"ResourceNotFoundException", "StateMachineDoesNotExist"}),
+			Hydrate:           getAwsStepFunctionsStateMachine,
```
Comment (bigdatasourav) [inline aws/table_aws_sfn_state_machine.go:20]:
```suggestion
			Hydrate:           getStepFunctionsStateMachine,
```

File: aws/table_aws_sfn_state_machine.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+		},
+		List: &plugin.ListConfig{
+			Hydrate: listAwsStepFunctionsStateManchines,
```
Comment (bigdatasourav) [inline aws/table_aws_sfn_state_machine.go:23]:
```suggestion
			Hydrate: listStepFunctionsStateManchines,
```

File: aws/table_aws_sfn_state_machine.go [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+				Name:        "title",
+				Description: resourceInterfaceDescription("title"),
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Name"),
+			},
+			{
+				Name:        "akas",
+				Description: resourceInterfaceDescription("akas"),
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("StateMachineArn").Transform(arnToAkas),
```
Comment (bigdatasourav) [inline aws/table_aws_sfn_state_machine.go:90]: use EnsureStringArray

File: aws/table_aws_sfn_state_machine.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+		func(page *sfn.ListStateMachinesOutput, isLast bool) bool {
+			for _, stateMachine := range page.StateMachines {
+				d.StreamListItem(ctx, stateMachine)
+			}
+			return !isLast
+		},
+	)
+	if err != nil {
+		plugin.Logger(ctx).Error("listAwsStepFunctionsStateManchines", "Error", err)
+		return nil, nil
```
Comment (bigdatasourav) [inline aws/table_aws_sfn_state_machine.go:117]:
```suggestion
		return nil, err
```

File: aws/table_aws_sfn_state_machine.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	return nil, err
```
Comment (bigdatasourav) [inline aws/table_aws_sfn_state_machine.go:120]:
```suggestion
	return nil, nil
```

File: docs/tables/aws_sfn_state_machine.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
+# Table: aws_sns_topic_subscription
```
Comment (bigdatasourav) [inline docs/tables/aws_sfn_state_machine.md:1]:
```suggestion
# Table: aws_sfn_state_machine
```

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Pull request merged**